rw/internal/platform/rates/rateapi/client.go

```diff
@@ -0,0 +1,119 @@
+package rateapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+)
+
+const (
+	operation = "usd to uah rate"
+)
+
+const (
+	usd = "USD"
+	uah = "UAH"
+)
+
+// usdUahResponse represents exchange rate API's response
+// ConversionRate is how much 1 USD is worth in a UAH.
+// Struct also contains some meta fields which can be useful in long run,
+// such as UpdatedAt and TargetCode.
+type usdUahResponse struct {
+	ConversionRate float32 `json:"conversion_rate"`
+	BaseCode       string  `json:"base_code"`
+	TargetCode     string  `json:"target_code"`
+	UpdatedAt      int     `json:"time_last_update_unix"`
+}
+
+// NewClient initializes new Client with parameters provided.
+// NOTE: neither of arguments can't be empty, because in that case
+// client will inevitably fail in the future.
+func NewClient(token, url string) *Client {
+	return &Client{
+		token: token,
+		url:   url,
+	}
+}
+
+//go:generate mockgen -destination=./mocks/mock_converter.go -package=mocks . Converter
+type Converter interface {
+	Convert(ctx context.Context) (float32, error)
+}
+
+// Client struct represents exchange rate API client.
+// Note: url should be a base url for  the service, not full url.
+type Client struct {
+	token string
+	next  Converter
+	url   string
+}
+
+// Convert method converts 1 USD to UAH accordingly to the
+// latest exchange rate. It's a handly wrapper around internal
+// getRate() function.
+func (c *Client) Convert(ctx context.Context) (float32, error) {
+	res := new(usdUahResponse)
+	err := c.getRate(ctx, res, usd, uah)
+	if err == nil {
+		return res.ConversionRate, nil
+	}
+
+	if c.next == nil {
+		return 0, fmt.Errorf("%s: %w", operation, err)
+	}
+
+	chainedRes, chainedErr := c.next.Convert(ctx)
+	if chainedErr != nil {
+		return 0, fmt.Errorf("%w: %w", err, chainedErr)
+	}
+
+	return chainedRes, nil
+}
+
+func (c *Client) SetNext(next Converter) {
+	c.next = next
+}
+
+// getRate method is used to query how much **from** currency is worth
+// in **to**  currency. response should be a pointer to the API response.
+// It's needed because API returns different responses for differect currency pairs.
+func (c *Client) getRate(
+	ctx context.Context,
+	response any,
+	from string,
+	to string,
+) error {
+	url, err := url.Parse(fmt.Sprintf("%s/%s/pair/%s/%s", c.url, c.token, from, to))
```

Comment:
Can we use here ` url.Values{}` structure 
for example 

`func ExampleValues() {
	v := url.Values{}
	v.Set("name", "Ava")
	v.Add("friend", "Jess")
	v.Add("friend", "Sarah")
	v.Add("friend", "Zoe")
	// v.Encode() == "name=Ava&friend=Jess&friend=Sarah&friend=Zoe"
	fmt.Println(v.Get("name"))
	fmt.Println(v.Get("friend"))
	fmt.Println(v["friend"])
	// Output:
	// Ava
	// Jess
	// [Jess Sarah Zoe]
}
`

Reply:
@KushnerykPavel `url.Values` constructs query parameters, but I have URL path, not query parameters, am I wrong?
https://pkg.go.dev/net/url#Values

Reply:
@KushnerykPavel I think I can parse base URL once at a creation time. Then I can use `JoinPath` function, I think it'd be better. What do you think?
https://pkg.go.dev/net/url#URL.JoinPath